### PR TITLE
volume resize: Fix issue with volume resize on VMWare (deploy as-is templates)

### PR DESCRIPTION
**Pull request by Pearl1594**
### Description

This PR fixes the issue pertaining to volume resize on VMWare for deploy as-is templates. VMware deploy as-is templates are those that are deployed as per the specification in the imported OVF. Hence override root disk size will not be adhered to for such templates. Moreover, when we deploy VMs in stopped state and resize the volume, the root disk doesn't get resized but the volume size is merely updated in the DB.
This PR also includes the following (for deploy as-is templates):
- Disables overriding root disk size during VM deployment on the UI
- Disables selection of compute offerings with root disk size specified, at the time of deployment
- Provided users with the option to deploy VM is stopped state via UI (so as to give an option to users to resize the volumes before starting the VM)

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [X] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

#### Prior to the fix:
Deployed a VM in stopped state, resized volume from 3GB to 8GB
image

While the UI shows that the Volume is resized,  the root volume inside the VM doesn't seem to have gotten resized:
image

#### Post fix
Deployed a VM in stopped state, resized the volume from 3GB to 8GB
image

image

Notice that the volume size as shown in the UI complies with the value inside the VM
Or verify the same on vcenter

### CODE REVIEW

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Review by nvazquez — COMMENTED** [review]
Looks good - perhaps also adding a check on the `resizevolume` and fail with a message when the volume is created from a deploy-as-is template?

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 1]
```diff
@@ -2072,6 +2072,8 @@ protected StartAnswer execute(StartCommand cmd) {
             //
             for (DiskTO vol : sortedDisks) {
                 if (vol.getType() == Volume.Type.ISO || deployAsIs && vol.getType() == Volume.Type.ROOT) {
+                    rootDiskTO = vol;
+                    resizeRootDiskOnVMStart(vmMo, rootDiskTO, hyperHost, context);
```
Comment (sureshanaparti) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:2076]: @Pearl1594 is this required for ISO as well?

Comment (Pearl1594) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:2080]: Thanks @sureshanaparti It's not required for ISO , have made the necessary change.

**Comment by Pearl1594** [comment]
> Looks good - perhaps also adding a check on the `resizevolume` and fail with a message when the volume is created from a deploy-as-is template?

@nvazquez added checks to see if a compute offering with root disk size mentioned is used / overriding root disk size is attempted when deploying a vm with deploy as-is template and will accordingly throw an exception.

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test matrix

**Comment by Pearl1594** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by Pearl1594** [comment]
@blueorangutan test centos7 vmware-67u3

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 2]
```diff
@@ -5266,6 +5270,16 @@ public UserVm createVirtualMachine(DeployVMCmd cmd) throws InsufficientCapacityE
             throw new InvalidParameterValueException("Unable to use template " + templateId);
         }
 
+        ServiceOfferingJoinVO svcOffering = serviceOfferingJoinDao.findById(serviceOfferingId);
+
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:5274]: @Pearl1594 is it possible to move all the checks after `template.isDeployAsIs()` to a separate if stmt (instead calling "template.isDeployAsIs()" multiple times)

Comment (Pearl1594) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:5274]: @sureshanaparti done

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test centos7 vmware-67u3

File: ui/public/locales/en.json [hunk 3]
```diff
@@ -1986,6 +1987,7 @@
 "label.start.rolling.maintenance": "Start Rolling Maintenance",
 "label.start.rolling.maintenance.payload": "Payload",
 "label.start.vlan": "Start VLAN",
+"label.start.vm": "Start VM",
```
Comment (nvazquez) [inline ui/public/locales/en.json:1990]: May not be needed, can we use the existing 'label.action.start.instance' instead?

File: ui/public/locales/en.json [hunk 4]
```diff
@@ -1807,6 +1807,7 @@
 "label.rolename": "Role",
 "label.roles": "Roles",
 "label.roletype": "Role Type",
+"label.rootdisksize": "Root disk size (GB)",
```
Comment (nvazquez) [inline ui/public/locales/en.json:1810]: Also may not be needed, can we use 'label.root.disk.size' instead? Is this one being used?

File: ui/src/views/compute/DeployVM.vue [hunk 5]
```diff
@@ -543,6 +545,9 @@
                         :options="keyboardSelectOptions"
                       ></a-select>
                     </a-form-item>
+                    <a-form-item :label="$t('label.start.vm')">
```
Comment (nvazquez) [inline ui/src/views/compute/DeployVM.vue:548]: As above: what about 'label.action.start.instance' here?

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java (hunk 2 above)
Comment (sureshanaparti) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:5274]: checked, looks good now.

File: ui/src/views/compute/DeployVM.vue (hunk 5 above)
Comment (Pearl1594) [reply ui/src/views/compute/DeployVM.vue]: done @nvazquez

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test matrix

**Review by nvazquez — APPROVED** [review]

**Review by sureshanaparti — APPROVED** [review]
code LGTM

**Comment by rhtyd** [comment]
Tests LGTM

**Comment by rhtyd** [comment]
With @shwstppr to manually test it

**Review by shwstppr — APPROVED** [review]
Tested on VMware 65u2 env. Works fine.
UI disables Root disk resize option for newly registered templates unlike earlier where it was shown but not honoured.
Earlier
Screenshot from 2021-03-26 16-53-06
Now
Screenshot from 2021-03-26 16-54-29

- Root volume resize of stopped VM deployed with a newly registered template (deployasis) - **Successful**
- Root volume resize of stopped VM deployed with an ISO - **Successfull**

**Review by shwstppr — CHANGES_REQUESTED** [review]
While volume resize behaviour works for most cases CKS cluster noderootdisksize is not working as expected. Will need similar changes from https://github.com/apache/cloudstack/pull/4329/commits/f5e866cb1f2a21e345f0dc1d6b6061960176d4df

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test centos7 vmware-67u3

**Review by shwstppr — APPROVED** [review]
LGTM. Tested CKS node root disk resize with latest changes

**Pull request merged**